### PR TITLE
Add TO Go cdns/capacity

### LINKED ISSUE #3780: Rewrite /cdns/capacity to Go
## I'm submitting an ...
improvement request (usability, performance, tech debt, etc.)

## Traffic Control components affected ...
- Documentation (possibly)
- Traffic Ops

## Current behavior:
`/cdns/capacity` is handled by Perl

## Expected / new behavior:
`/cdns/capacity` should be handled by Go

### Minimum API Version
1.1

### Request Methods
- [ ] `GET`

## Minimal reproduction of the problem with instructions:
N/A

## Anything Else

PR in progress: #2306

### PR DESCRIPTION
**Pull request by rob05c**
#### What does this PR do?

Add TO Go cdns/capacity. Closes #3780 

#### Which TC components are affected by this PR?

- [ ] Documentation
- [ ] Grove
- [ ] Traffic Analytics
- [ ] Traffic Monitor
- [x] Traffic Ops
- [ ] Traffic Ops ORT
- [ ] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [ ] Traffic Vault
- [ ] Other _________

#### What is the best way to verify this PR?

`GET /api/1.2/cdns/capacity`

#### Check all that apply

- [ ] This PR includes tests
- [ ] This PR includes documentation updates
- [ ] This PR includes an update to CHANGELOG.md
- [ ] This PR includes all required license headers
- [ ] This PR includes a database migration (ensure that migration sequence is correct)
- [ ] This PR fixes a serious security flaw. Read more: [www.apache.org/security](http://www.apache.org/security/)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1625/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1654/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1763/
Test PASSed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1776/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/1973/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2040/
Test PASSed.

**Comment by rob05c** [comment]
#### What does this PR do?

Add TO Go cdns/capacity

#### Which TC components are affected by this PR?

- [ ] Documentation
- [ ] Grove
- [ ] Traffic Analytics
- [ ] Traffic Monitor
- [x] Traffic Ops
- [ ] Traffic Ops ORT
- [ ] Traffic Portal
- [ ] Traffic Router
- [ ] Traffic Stats
- [ ] Traffic Vault
- [ ] Other _________

#### What is the best way to verify this PR?

`GET /api/1.2/cdns/capacity`

#### Check all that apply

- [ ] This PR includes tests
- [ ] This PR includes documentation updates
- [ ] This PR includes an update to CHANGELOG.md
- [ ] This PR includes all required license headers
- [ ] This PR includes a database migration (ensure that migration sequence is correct)
- [ ] This PR fixes a serious security flaw. Read more: [www.apache.org/security](http://www.apache.org/security/)

**Comment by rob05c** [comment]
Rebased, should be good to merge.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2436/
Test PASSed.

**Comment by mitchell852** [comment]
Is there any sort of API test that can be added? Maybe one to check the response structure or something?

**Comment by rob05c** [comment]
@mitchell852 Not until the "API test" framework supports Monitors.

**Comment by ocket8888** [comment]
This endpoint has already been rewritten to go, so I think this PR should be closed.

**Comment by rob05c** [comment]
This hasn't been rewritten in Go, except for this PR. It's still proxtying to Perl https://github.com/apache/trafficcontrol/blob/master/traffic_ops/traffic_ops_golang/routing/routes.go#L123

**Comment by ocket8888** [comment]
Oh, lol, my bad. I forget that some of those routes exist for explicit routing to the proxy. If you wanna rebase for conflicts, I'd be happy to review this.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4142/
Test PASSed.

File: traffic_ops/traffic_ops_golang/cdn/capacity.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+	MaxKBPS []CacheStatData `json:"maxKbps"`
+}
+
+type CacheStatData struct {
+	Value float64 `json:"value,string"`
+}
+
+// getCacheStats gets the cache stats from the given monitor. It takes stats, a slice of stat names; and cacheStats, an object to deserialize stats into. The cacheStats type must be of the form struct {caches map[tc.CacheName]struct{statName []struct{value float64}}} with the desired stats, with appropriate member names or tags.
+func getCacheStats(monitorFQDN string, client *http.Client, stats []string, cacheStats interface{}) error {
+	path := `/publish/CacheStats`
```
Comment (mitchell852) [inline traffic_ops/traffic_ops_golang/cdn/capacity.go]: am i wrong or should this just be `/CacheStats`?

File: traffic_ops/traffic_ops_golang/cdn/capacity.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+	if len(stats) > 0 {
+		path += `?stats=` + strings.Join(stats, `,`)
+	}
+	resp, err := client.Get("http://" + monitorFQDN + path)
+	if err != nil {
+		return errors.New("getting CacheStats from Monitor '" + monitorFQDN + "': " + err.Error())
+	}
+	defer resp.Body.Close()
+	if err := json.NewDecoder(resp.Body).Decode(cacheStats); err != nil {
+		return errors.New("decoding CacheStats from monitor '" + monitorFQDN + "': " + err.Error())
```
Comment (mitchell852) [inline traffic_ops/traffic_ops_golang/cdn/capacity.go]: i changed it to `/CacheStats` and now i'm getting this error when hitting an online monitor:

`ERROR: api.go:122: 2019-08-16T15:14:35.164661-06:00: 127.0.0.1:65348 getting cache stats for CDN 'cdn-1' monitor 'cdn-trafficmonitor-foo.bar.net:80': decoding CacheStats from monitor 'cdn-trafficmonitor-foo.bar.net:80': invalid character '<' looking for beginning of value
`

Comment (mitchell852) [reply traffic_ops/traffic_ops_golang/cdn/capacity.go]: nevermind. publish is the json version. duh.

File: traffic_ops/traffic_ops_golang/cdn/capacity.go (hunk 1 above)
Comment (mitchell852) [reply traffic_ops/traffic_ops_golang/cdn/capacity.go]: nevermind

File: traffic_ops/traffic_ops_golang/cdn/capacity.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+	thresholds, err := getEdgeProfileHealthThresholdBandwidth(tx)
+	if err != nil {
+		return CapacityResp{}, errors.New("getting profile thresholds: " + err.Error())
+	}
+
+	cap := CapData{}
+	for cdn, monitorFQDN := range monitors {
+		crStates, err := getCRStates(monitorFQDN, client)
+		// TODO on err, try another online monitor
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/cdn/capacity.go]: I think this was written to behave exactly like Perl. But that does seem like a pretty safe change, I can go ahead and make it now if you want

File: traffic_ops/traffic_ops_golang/cdn/capacity.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+		return errors.New("getting CacheStats from Monitor '" + monitorFQDN + "': " + err.Error())
+	}
+	defer resp.Body.Close()
+	if err := json.NewDecoder(resp.Body).Decode(cacheStats); err != nil {
+		return errors.New("decoding CacheStats from monitor '" + monitorFQDN + "': " + err.Error())
+	}
+	return nil
+}
+
+func getMonitorForwardProxy(tx *sql.Tx) (string, error) {
```
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/cdn/capacity.go]: It wasn't, removed.

File: traffic_ops/traffic_ops_golang/cdn/capacity.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rob05c) [reply traffic_ops/traffic_ops_golang/cdn/capacity.go]: Ok, I changed it to try the next monitor if one fails, and only error if all monitors for a CDN fail.

I do think an ISE is appropriate if that happens. It is a server error, returning a capacity of 0 wouldn't be correct.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4178/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4179/
Test FAILed.

**Pull request merged**